lib/midi_generator.rb

```diff
@@ -0,0 +1,135 @@
+# Converts Inevitable Cacophony internal note representation
+# into MIDI messages usable by an external synthesizer.
+# Based on examples in the `midilib` gem.
+
+require 'midilib/sequence'
+require 'midilib/consts'
+
+class MidiGenerator
+
+        # Middle A in MIDI
+        MIDI_TONIC = 69
+
+        # Add a phrase to the MIDI output we will generate.
+        def add_phrase(phrase)
+                @phrases ||= []
+                @phrases << phrase
+        end
+
+        # @return [Midi::Track] Notes to be output to MIDI; mainly for testing.
+        def notes_track
+                build_notes_track(sequence, @phrases)
+        end
+
+        # Write MIDI output to the given stream.
+        def write(io)
+                sequence.tracks << notes_track
+
+                # Buffer output so we can send to stdout.
+                buffer = StringIO.new
+                sequence.write(buffer)
+
+                io.write(buffer.string)
+        end
+
+        # Create an array with frequencies in Hertz for each
+        # MIDI note in the given octave structure.
+        #
+        # @param octave_structure [OctaveStructure]
+        # @param tonic [Integer] The tonic frequency in Hertz.
+        #                        This will correspond to Cacophony frequency 1,
+        #                        and MIDI pitch 69
+        def frequency_table(octave_structure, tonic)
```

Comment:
Why is this code not merged with, or at least closer to, the [`#midi_index`](https://github.com/isikyus/inevitable-cacophony/pull/9/files#diff-fc3419ccd3d92a8c15b2df1e6887d1d1R130) method that does the conversion the other way.

Reply:
The problem here is that `#midi_index` is approximate, while the algorithm here is exact. So this code has to be the authoritative version. But it needs the OctaveStructure object, which we don't have in the `#midi_index` method — only an individual note.

Passing in OctaveStructure to `#midi_index` is tricky, since it's global data that doesn't seem relevant at the level of that method — it'd be better if we had a more detailed "note" or "pitch" object in that method that already knew where in the overall octave structure it sat.

Reply:
These two methods actually use completely different approaches at the moment. We probably need to use one or the other depending on how much retuning we need:

* If `notes_per_octave` is 12 or fewer, we can fit our scale notes inside MIDI's standard octaves, so the pitches in the raw MIDI file are roughly where they should be.
  * If the notes are also spaced evenly enough, we can pick the closest MIDI pitch for each individual note. Many of Dwarf Fortress's scales actually _are_ subsets of 12TET, in which case we can use the exact MIDI note and re-tuning isn't necessary at all.
    This is the approach `#midi_index` currently uses.
  * If they aren't spaced that well — e.g. two notes a quarter-tone apart — we need to pick less-accurate MIDI indices for some notes to avoid data loss. Merge them, and no amount of retuning will get get those distinctions back.
    The code in this method will work for this, but it has the disadvantage of losing octave-for-octave compatibility unnecessarily.
* If there are more than 12 notes per octave, fitting the "chromaitc" notes into MIDI octaves is impossible, and we need to re-scale pretty much every single note. That's what this method currently does.

Reply:
To get enough information to make this decision, I've ended up requiring an OctaveStructure and tonic note to generate MIDI, like how we already need a tonic note to generate WAV.

Reply:
I also gave up on the idea of getting close --- only an "exact" match (within human perception limits) is worthwhile as otherwise it's still going to sound out of tune. Maybe not always -- think 12TET versus Just Intonation -- but good enough for now, particularly since it's only ever going to be approximate/imperfect.